### ISSUE TITLE
Integrate value function fitting, model learning and policy improvement

### ISSUE THREAD

**Issue by thiagopbueno** [thiagopbueno/model-aware-policy-optimization#12]
Integrate all optimization step ops for value function fitting, model learning, and policy improvement into `Policy. learn_on_batch()` or `Policy.compute_gradients()` and `Policy.apply_gradients`. If necessary use TensorFlow's `tf.control_dependencies` feature.

**Comment by thiagopbueno** [comment]
In order to integrate all learning procedures, should we:

1. extend the [Trainer](https://github.com/ray-project/ray/blob/54d5969ceacabad2741af278ef9ed4db0af66835/python/ray/rllib/agents/trainer.py#L264) class; or
2. call the helper function [build_trainer](https://github.com/ray-project/ray/blob/54d5969ceacabad2741af278ef9ed4db0af66835/python/ray/rllib/agents/trainer_template.py#L14) ?

**Comment by angelolovatto** [comment]
> Integrate all optimization step ops for value function fitting, model learning, and policy improvement into `Policy. learn_on_batch()` or `Policy.compute_gradients()` and `Policy.apply_gradients`. If necessary use TensorFlow's `tf.control_dependencies` feature.

If we plan on doing separate training loops for each component (e.g. fitting the value function several times while only doing one policy gradient step), I think it makes more sense to implement `Policy.learn_on_batch()` because we wouldn't be taking only one gradient step (putting it another way, I don't see how `Policy.compute_gradients()` would summarize all of these separate training procedures).
